### DISCUSSION TITLE
How to set a 48 hour timer?

### DISCUSSION THREAD

**Discussion by 49studebaker** [arendst/Tasmota#18795] [Q&A]
I would like to turn ON my smart plug every 48 hours. I looked thru the list of tasmota commands and “RuleTimer” has a maximum of 65535 seconds(18.2 hours). How could I set a 48 hour timer? After the plug is turned ON, I will use PulseTime 130, to turn it OFF after 30 seconds.

**Comment by sfromis** [comment]
One way would be to use the "Configure Timers" dialog, to have a timer popping daily, and then have a variable to check for every 2nd time, like initialized to 1, and multiplying by -1 every 24 hours. If it is then > 0, 48 hours passed.

**Reply by 49studebaker** [reply]
When I plug in my smart plug, I would like Power1 to remain OFF. If there is an electrical storm, the device would turn on/off, as the AC power cycles on/off.

Daily timer is set to turn on rule.

The code sets var1 to -1. Why is var1  1?


Current code:

Rule1
ON System#Init DO Backlog Time 1672635600; Var1 1; Var2 0 ENDON
ON Clock#Timer=1 DO Backlog Mult1 -1; Var2 %Var1% ENDON
ON Var2#Data>0 DO Power1 1 ENDON

Log:

00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F9, Count 771
00:00:00.055 SER: Set to 8E1 4800 bit/s
00:00:00.056 QPC: Count 1
00:00:00.073 NRG: Init driver 2
00:00:00.084 Project tasmota - Tasmota Version 12.5.0(tasmota)-2_7_4_9(2023-04-17T08:03:54)
00:00:00.088 SER: Set to 8E1 4800 bit/s
00:00:00.091 SNS: Hardware Serial
00:00:00.154 RUL: SYSTEM#INIT performs "Backlog Time 1672635600; Var1 1; Var2 0"
00:00:00.003 RSL: RESULT = {"Time":"2023-01-02T00:00:00","Epoch":1672635600}
00:00:00.357 WIF: Connecting to AP1 CH iPhone Channel 6 BSSId EE:D6:A8:B9:5B:AA in mode 11n as tasmota-6632D7-4823...
00:00:00.412 RSL: RESULT = {"Var1":"1"}
00:00:00.619 RSL: RESULT = {"Var2":"0"}
00:00:05.918 QPC: Reset
00:00:13.777 WIF: Connect failed with AP timeout
00:00:14.115 WIF: Connecting to AP1 CH iPhone in mode 11n as tasmota-6632D7-4823...
00:00:22.694 WIF: Connected
00:00:22.929 HTP: Web server active on tasmota-6632D7-4823 with IP address 172.20.10.2
00:00:23.956 RSL: INFO1 = {"Info1":{"Module":"Sonoff S31","Version":"12.5.0(tasmota)","FallbackTopic":"cmnd/DVES_6632D7_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:23.966 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-6632D7-4823","IPAddress":"172.20.10.2"}}
00:00:23.976 RSL: INFO3 = {"Info3":{"RestartReason":"Power On","BootCount":100}}
00:00:23.985 RSL: RESULT = {"POWER":"OFF"}
00:00:23.987 RSL: POWER = OFF
00:00:28.634 RSL: STATE = {"Time":"2023-01-02T00:00:28","Uptime":"0T00:00:29","UptimeSec":29,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"CH iPhone","BSSId":"3E:F1:8F:F6:40:34","Channel":6,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:23"}}
00:00:28.647 RSL: SENSOR = {"Time":"2023-01-02T00:00:28","ENERGY":{"TotalStartTime":"2023-05-25T22:26:44","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"ApparentPower":0,"ReactivePower":0,"Factor":0.00,"Voltage":0,"Current":0.000}}
00:03:52.818 CMD: Time
00:03:52.823 RSL: RESULT = {"Time":"2023-01-02T00:03:52"}
00:05:00.668 RUL: CLOCK#Timer=1 performs "Backlog Mult1 -1; Var2 1"
00:05:00.688 RSL: RESULT = {"Mult1":"-1.000"}
00:05:00.889 RSL: RESULT = {"Var2":"1"}
00:05:00.905 RUL: VAR2#DATA>0 performs "Power1 1"
00:05:00.910 RSL: RESULT = {"POWER":"ON"}
00:05:00.913 RSL: POWER = ON
00:05:28.621 RSL: STATE = {"Time":"2023-01-02T00:05:28","Uptime":"0T00:05:29","UptimeSec":329,"Heap":22,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"ON","Wifi":{"AP":1,"SSId":"CH iPhone","BSSId":"3E:F1:8F:F6:40:34","Channel":6,"Mode":"11n","RSSI":100,"Signal":-49,"LinkCount":1,"Downtime":"0T00:00:23"}}
00:05:28.632 RSL: SENSOR = {"Time":"2023-01-02T00:05:28","ENERGY":{"TotalStartTime":"2023-05-25T22:26:44","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"ApparentPower":0,"ReactivePower":0,"Factor":0.00,"Voltage":121,"Current":0.000}}
00:05:33.004 RSL: RESULT = {"POWER":"OFF"}
00:05:33.006 RSL: POWER = OFF

**Reply by sfromis** [reply]
Variables are expanded before the rule starts. Thus, you cannot first modify and then use a var in the same rule.

And correct trigger would be: `ON Var2#State>0 DO`

One trick to avoid two rules could be: `Scale1 %var1%,1,0,0,1`
This flips the value of var1 between 0 and 1 each time.

**Reply by 49studebaker** [reply]
Is this correct?

Rule1
ON System#Init DO Backlog Time 1672635600; Var1 0 ENDON
ON Clock#Timer=1 DO Scale1 %var1%,1,0,0,1 ENDON
ON Var1#State>0 DO Power1 1 ENDON

**Comment by 49studebaker** [comment]
I think this might work:

```
Var1 0.000

Rule1
ON Rules#Timer=1 DO Backlog Add1 1; RuleTimer 43200; Var1 ENDON
ON Var1#Data=4.000 DO Backlog Var1 0.000;  Power1 1 ENDON

Rule1 1

RuleTimer1 43200
```

**Reply by sfromis** [reply]
The downside of that technique is that it is highly fragile in case of the device having an outage for some reason. That's why I instead suggested using configure timers.

Instead of `ON Var1#Data=4.000 DO Backlog Var1 0.000;` better use `ON Var1#Data==4 DO Backlog Var1 0;`, doing numerical compare avoids that mess with decimals. You could also use `CalcRes 0` to have zero decimals.

**Comment by 49studebaker** [comment]
I would like it to work without being dependent on Wi-Fi. Worse case, if there is a power outage, it will turn ON after 96hrs and then every 48hrs going forward. Will any problems occur if Tasmota doesn’t connect to Wi-Fi for months?

Rule1
ON System#Init DO Backlog Var1 0; RuleTimer 43200 ENDON
ON Rules#Timer=1 DO Backlog Add1 1; RuleTimer 43200; Var1 ENDON
ON Var1#Data==4 DO Backlog Var1 0;  Power1 1 ENDON

**Reply by sfromis** [reply]
Tasmota is not designed to work seamlessly without a real time source, which by default is NTP via Wifi, and may reboot while trying to get a connection. And do not expect the MCU timing to be accurate without resync. And as already mentioned, using ruletimer+vars to manage the cycle is highly fragile to "things happening".

**Reply by 49studebaker** [reply]
> Tasmota is not designed to work seamlessly without a real time source, which by default is NTP via Wifi, and may reboot while trying to get a connection. And do not expect the MCU timing to be accurate without resync. And as already mentioned, using ruletimer+vars to manage the cycle is highly fragile to "things happening".


If I turn off Wi-Fi, will it prevent tasmota from rebooting because tasmota can not acquire a connection?

Rule3 
ON System#Init DO Wifi 0 ENDON
ON Button1#state=3 DO Wifi 1 ENDON

How much does MCU timing drift every month?

Thanks.

**Reply by barbudor** [reply]
Assuming the oscillator is 100ppm (average for cheap quality), that would mean up to 8 sec per day
But better test that for yourself

**Reply by 49studebaker** [reply]
ON System#Init DO Wifi 0 ENDON


My phone detects the "esp8266" Wi-Fi network. After connecting to the esp Wi-Fi network, I am unable to access the tasmota home screen. Is tasmota turning ON Wi-Fi after the Wifi 0 command is executed?

If I hold the s31 power button, I can connect to the s31 Wi-Fi network and access the tasmota home screen.

If I run the  wifi 0  command on the tasmota console, my phone does not detect the esp wifi network.

Should I use a different trigger to turn off wifi on startup?

**Reply by sfromis** [reply]
Tasmota is not "turning on" Wifi, but it may not be turning it off when you disconnect.

**Reply by pkkrusty** [reply]
If time of day and drift don't matter than much, you can set time with `Time` command on startup to stop the time code from searching for NTP servers and preventing triggers from happening.

**Reply by pkkrusty** [reply]
Another option is to flash with ALWAYS_AP hack that I use for remote locations with no wifi. Keeps AP mode always active while allowing full control and prevents restart due to no wifi. Search forums for "ALWAYS_AP"

**Reply by 49studebaker** [reply]
> ON System#Init DO Wifi 0 ENDON
> 
> 
> My phone detects the "esp8266" Wi-Fi network. After connecting to the esp Wi-Fi network, I am unable to access the tasmota home screen. Is tasmota turning ON Wi-Fi after the Wifi 0 command is executed?
> 
> If I hold the s31 power button, I can connect to the s31 Wi-Fi network and access the tasmota home screen.
> 
> If I run the  wifi 0  command on the tasmota console, my phone does not detect the esp wifi network.
> 
> Should I use a different trigger to turn off wifi on startup?


Previously, when I restarted the smart plug, the device would appear in my phone's list of Wi-Fi networks. I could not access the device's Home Screen without holding down the power button for X seconds.

I changed   System#Init  to  System#Boot

When I restart my smart plug, the device does NOT appear in my phone's list of Wi-Fi networks. When I hold down the power button for X seconds, I am able to connect to the smart plug and access the Home Screen.

**Reply by 49studebaker** [reply]
How would I modify this code so that it toggles Wi-Fi on/off when the button is pressed and held?

Rule3 
ON System#Boot DO Wifi 0 ENDON
ON Button1#state=3 DO Wifi 1 ENDON

**Reply by barbudor** [reply]
There is no toggle feature on the Wifi command so you need to be a bit tricky
```
Rule3
  ON System#Boot DO Wifi 0 ENDON
  ON Wifi=ON DO Var1 0 ENDON
  ON Wifi=OFF DO Var1 1 ENDON
  ON Button1#state=3 DO Wifi %var1% ENDON
```

**Comment by 49studebaker** [comment]
Rule1
ON Clock#Timer=1 DO Backlog Mult1 -1; Var1 ENDON
ON Var1#Data>0 DO Backlog Var1 1; Power1 1 ENDON

How would I first initialize Var1 to 1 without executing ON Var1#Data>0

**Reply by barbudor** [reply]
The same way you were initializing it to 0 in the previous version

Don't forget that Tasmota may automatically restart when wifi is lost for more than xx minutes which would reset your rule
If you change it only once a day, you would be safer with a mem than a var. You can then initialize the mem once in the console, not from a boot rule.

**Reply by 49studebaker** [reply]
I think I tried that but it keeps executing “ON Var1#Data>0 DO Backlog Var1 1; Power1 1 ENDON”. On init Var1 is equal to 1, which is greater than zero. It executes that line of code.

Rule1
ON System#Init DO Backlog Time [1672534800](tel:1672534800); Var1 1 ENDON
ON Clock#Timer=1 DO Backlog Mult1 -1; Var1 ENDON
ON Var1#Data>0 DO Backlog Var1 1; Power1 1 ENDON

**Reply by barbudor** [reply]
Haven't you set the rule to 'once' mode so it is executed only once?  (see Rule<x> 5)

**Comment by btsimonh** [comment]
It may be worth looking at deepsleeptime.
I was just playing with this on ESP32, and it can sleep the ESP for up to 10 years.
The time the ESP is awake can be controlled with teleperiod.
so `teleperiod 30`
then `deepsleeptime 172800`
The question will be does it open the relay when it goes to sleep?

**Reply by stefanbode** [reply]
On deepsleep you can define action or better rules that execute before deepsleep. Therefore switching a relay before deepsleep is possible.
